proxy/src/project.rs

```diff
@@ -80,14 +86,154 @@ pub enum Registration {
+                "credential.helper=!f() {{ test \"$1\" = get && echo \"password={}\"; }}; f",
+                pass
+            ),
+        }
+    }
+}
+
+/// The data necessary for checking out a project.
+pub struct Checkout<P, ST>
+where
+    P: AsRef<path::Path>,
+{
+    /// The credential helper.
+    credential: Credential,
+    /// The project.
+    project: coco::Project<ST>,
+    /// The path on the filesystem where we're going to checkout to.
+    path: P,
+    /// The `PATH` environment variable to be used for the checkout. It is safe to leave this
+    /// `None` when executing the application for real. However, if we want to run an integration
+    /// test we need to tell say where the `git-rad-remote` helper can be found.
+    bin_path: Option<ffi::OsString>,
+}
+
+impl<P, ST> Checkout<P, ST>
+where
+    P: AsRef<path::Path>,
+    ST: Clone,
+{
+    /// Create a new `Checkout` with the mock `Credential::Password` helper.
+    pub fn new<Bin>(project: coco::Project<ST>, path: P, bin_path: Bin) -> Self
+    where
+        Bin: Into<Option<ffi::OsString>>,
+    {
+        Self {
+            // TODO(rudolfs): we'll have to figure out how to pass the secret
+            // key to git in a safe manner. As it is now it could be sniffed
+            // out from the process list while the user is doing a clone.
+            //
+            // How will we get ahold on the secret key here?
+            credential: Credential::Password("radicle-upstream".to_owned()),
+            project,
+            path,
+            bin_path: bin_path.into(),
+        }
+    }
+
+    /// Checkout a working copy of a [`Project`].
+    ///
+    /// NOTE: `RAD_HOME` should be expected to be set if using a custom root for
+    /// [`librad::paths::Paths`]. If it is not set the underlying binary will delegate to the
+    /// `ProjectDirs` setup of the `Paths`.
+    ///
+    /// # Errors
+    ///
+    ///   * We couldn't resolve the executable path.
+    ///   * The checkout process failed.
+    pub fn run(self) -> Result<path::PathBuf, error::Error> {
+        let bin_path = match self.bin_path {
+            Some(path) => Ok(path),
+            None => Self::default_bin_path(),
+        }?;
+
+        // Check if the path provided ends in the 'directory_name' provided. If not we create the
+        // full path to that name.
+        let path = &self.path.as_ref();
+        let project_path = if let Some(destination) = path.components().next_back() {
+            let destination: &std::ffi::OsStr = destination.as_ref();
+            let project_name = self.project.name().to_string();
+            let name: &std::ffi::OsStr = project_name.as_ref();
+            if destination == name {
+                path.to_path_buf()
+            } else {
+                path.join(name)
+            }
+        } else {
+            path.join(&self.project.name().to_string())
+        };
+
+        let mut child_process = Command::new("git")
+            .arg("-c")
```

Comment:
It might help readability if we used the long names for these params, now that I think about it

Reply:
I can't seem to get the credential helper to work with the `config` flag. Wanna show me how it's done?

Reply:
Ah maybe not this one, since the `key=value` is above in the `Credential`.

But `--origin` and `--branch` would be better :)

Reply:
Good point!

https://github.com/radicle-dev/radicle-upstream/pull/722/commits/4c2e8ae98de6c19c0c1597ef990fb12df0680721